types/wlr_keyboard_group.c

```diff
@@ -0,0 +1,325 @@
+	wl_list_for_each_safe(key, tmp, &group->keys, link) {
+		if (key->keycode != event->keycode) {
+			continue;
+		}
+		if (event->state == WLR_KEY_PRESSED) {
+			key->count++;
+			return;
+		}
+		if (event->state == WLR_KEY_RELEASED) {
+			key->count--;
+			if (key->count > 0) {
+				return;
+			}
+			wl_list_remove(&key->link);
+			free(key);
+		}
+		break;
+	}
+
+	if (event->state == WLR_KEY_PRESSED) {
+		struct keyboard_group_key *key =
+			calloc(1, sizeof(struct keyboard_group_key));
+		if (!key) {
+			wlr_log(WLR_ERROR, "Failed to allocate keyboard_group_key");
+			return;
+		}
+		key->keycode = event->keycode;
+		key->count = 1;
+		wl_list_insert(&group->keys, &key->link);
+	}
+
+	wlr_keyboard_notify_key(&group_device->keyboard->group->keyboard, data);
+	wlr_keyboard_group_led_update_from_xkb_state(group);
+}
+
+static void handle_keyboard_modifiers(struct wl_listener *listener,
+		void *data) {
+	// Sync the effective layout (group modifier) to all keyboards. The rest of
+	// the modifiers will be derived from the wlr_keyboard_group's key state
+	struct keyboard_group_device *group_device =
+		wl_container_of(listener, group_device, modifiers);
+	uint32_t group = group_device->keyboard->modifiers.group;
+
+	struct keyboard_group_device *device;
+	wl_list_for_each(device, &group_device->keyboard->group->devices, link) {
+		if (group != device->keyboard->modifiers.group) {
+			wlr_keyboard_notify_modifiers(device->keyboard,
+					device->keyboard->modifiers.depressed,
+					device->keyboard->modifiers.latched,
+					device->keyboard->modifiers.locked, group);
+			return;
+		}
+	}
+
+	struct wlr_keyboard *groupkb = &group_device->keyboard->group->keyboard;
+	wlr_keyboard_notify_modifiers(groupkb, groupkb->modifiers.depressed,
+			groupkb->modifiers.latched, groupkb->modifiers.locked, group);
+}
+
+static void handle_keyboard_keymap(struct wl_listener *listener, void *data) {
+	struct keyboard_group_device *group_device =
+		wl_container_of(listener, group_device, keymap);
+	struct wlr_keyboard *keyboard = group_device->keyboard;
+
+	if (!keymaps_match(keyboard->group->keyboard.keymap, keyboard->keymap)) {
+		struct keyboard_group_device *device;
+		wl_list_for_each(device, &keyboard->group->devices, link) {
+			if (!keymaps_match(keyboard->keymap, device->keyboard->keymap)) {
+				wlr_keyboard_set_keymap(device->keyboard, keyboard->keymap);
+				return;
+			}
+		}
+	}
+
+	wlr_keyboard_set_keymap(&keyboard->group->keyboard, keyboard->keymap);
+}
+
+static void handle_keyboard_repeat_info(struct wl_listener *listener,
+		void *data) {
+	struct keyboard_group_device *group_device =
+		wl_container_of(listener, group_device, repeat_info);
+	struct wlr_keyboard *keyboard = group_device->keyboard;
+
+	struct keyboard_group_device *device;
+	wl_list_for_each(device, &keyboard->group->devices, link) {
+		struct wlr_keyboard *devkb = device->keyboard;
+		if (devkb->repeat_info.rate != keyboard->repeat_info.rate ||
```

Comment:
This check should probably be in `wlr_keyboard_set_repeat_info` instead.

Reply:
This is used to sync the repeat info to the other keyboards in the group so I think it makes more sense here